src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -1125,6 +1125,9 @@ namespace QgsWms
     {
       mapSettings.setBackgroundColor( backgroundColor );
     }
+
+    // add context from project (global variables, ...)
+    mapSettings.setExpressionContext( mProject->createExpressionContext() );
```

Comment:
This may be done elsewhere, but I believe you also want a mapSettingsScope added here

Reply:
Hi @nyalldawson,

> This may be done elsewhere, but I believe you also want a mapSettingsScope added here

Could you elaborate a little bit more on that? Because a `mapSettingsScope` would contain variables related to  `QgsMapSettings` (like map rotation), but I don't see where it can be directly useful for the rendering step on the server side?

Thanks for your review!

Reply:
But in this case, the `mapSettings` variable is already configured with extent, size and so on. 

Then it's directly used by `QgsMapRendererParallelJob` or `QgsMapRendererCustomPainterJob` for the rendering: 

```
QgsMapRendererParallelJob renderJob( mapSettings );
```

So I don't see how the `mapSettingsScope` could be usfeul in this scenario? 

Sorry, it's probably obvious... :)

Reply:
If you had a rule (or data defined symbology) which relies on any of the variables provided through the map settings scope, then you need this scope added to the expression context or the expressions won't evaluate correctly. (This includes @map_scale, @map_extent, @map_rotation, etc).

Reply:
> then you need this scope added to the expression context or the expressions won't evaluate correctly. 

Of course, thank you for your explanation Nyall!

I updated the expression context and unit tests accordingly to check that `@map` variables are correctly used :).

Reply:
Perfect!